go/internal/tinder/driver_multi.go

```diff
@@ -48,126 +84,189 @@ func (md *MultiDriver) Advertise(ctx context.Context, ns string, opts ...p2p_dis
+	// We don't use a lock array because what we execute next is so cheap it's not worth the extra context switches that generates.
+	var wg sync.WaitGroup
+	wg.Add(ndrivers)
+	errReturn := make([]error, ndrivers)
+	durationReturn := make([]time.Duration, ndrivers)
+	for i := 0; i < ndrivers; i++ {
+		go func(j int) {
+			durationReturn[j], errReturn[j] = md.drivers[j].Advertise(ctx, ns, opts...)
+			wg.Done()
+		}(i)
+	}
+	wg.Wait()
+
+	for i, v := range errReturn {
+		// Check if at least one was successful
+		if v == nil {
+			// One was, fetch the longer ttl
+			bestDuration := durationReturn[i]
+			for j := i + 1; j < ndrivers; j++ {
+				durj := durationReturn[j]
+				if durj > bestDuration {
+					bestDuration = durj
+				}
+			}
+			if options.Ttl != 0 && bestDuration > options.Ttl {
+				return options.Ttl, nil
+			}
+			return bestDuration, nil
+		}
 	}
 
-	return options.Ttl, nil
+	return 0, multierr.Combine(errReturn...)
 }
 
-func (md *MultiDriver) advertise(ctx context.Context, d Driver, ns string, opts ...p2p_discovery.Option) {
-	go func() {
-		for {
-			ttl, err := d.Advertise(ctx, ns, opts...)
-			if err != nil {
-				if ctx.Err() != nil {
-					return
-				}
+const maxLimit = 1000
 
-				md.logger.Warn("failed to advertise",
-					zap.String("driver", d.Name()),
-					zap.String("key", ns),
-					zap.Error(err),
-				)
+func (md *MultiDriver) FindPeers(ctx context.Context, ns string, opts ...p2p_discovery.Option) (<-chan p2p_peer.AddrInfo, error) {
+	var options p2p_discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return nil, err
+	}
+	limit := options.Limit
+	if limit == 0 || limit > maxLimit {
+		limit = maxLimit
+	}
 
+	// Buffer to try to limit context switch while returning peers.
+	outPeers := make(chan p2p_peer.AddrInfo, 10)
+	ctx, cancelSearch := context.WithCancel(ctx)
+	// This wait group is used to close outPeers when we are finished.
+	var wg sync.WaitGroup
+	ndrivers := len(md.drivers)
+	wg.Add(ndrivers)
+	errReturn := make([]error, ndrivers)
+	// This wait group is used to know when to check for errors.
+	var errWg sync.WaitGroup
+	errWg.Add(ndrivers)
+	var alreadySeen []p2p_peer.ID
+	var seenLock sync.RWMutex
+	for i := 0; i < ndrivers; i++ {
+		go func(j int) {
+			defer wg.Done()
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			var inPeers <-chan p2p_peer.AddrInfo
+			inPeers, errReturn[j] = md.drivers[j].FindPeers(ctx, ns, opts...)
+			errWg.Done()
+		MainSearchLoop:
+			for v := range inPeers {
+				id := v.ID
+				seenLock.RLock()
+				oldlen := len(alreadySeen)
+				for _, w := range alreadySeen {
+					if w == id {
+						seenLock.RUnlock()
+						continue MainSearchLoop
+					}
+				}
+				seenLock.RUnlock()
+				seenLock.Lock()
+				// We need to check if newer peers were added while we were not locking.
+				curlen := len(alreadySeen)
+				if curlen > limit {
+					return
+				}
+				for oldlen < curlen {
+					if alreadySeen[oldlen] == id {
```

Comment:
don't really know idea here.
But because oldlen < curlen, and maybe alreadySeen[oldlen] != id.
but maybe id will happen in  alreadySeen[oldlen+1] to curlen-1 ?
Why don't use map here with key is id, value is bool? This will make code simpler than?